### PR TITLE
feat(gsd): codebase map — structural orientation for fresh agent contexts

### LINKED ISSUE #2229: feat: Codebase Map — structural orientation for fresh agent contexts
## Problem

Every fresh agent context pays an "exploration tax" — 10-30+ tool calls just to understand what exists and where things live. GSD's existing artifacts (PROJECT.md, KNOWLEDGE.md, DECISIONS.md) provide project-level context (what we're building, what we've learned, what we've decided), but none answer the structural question: *"what does this file do?"* or *"what's in this directory and why?"*

This leads to:
- Wasted context window on repeated filesystem exploration
- Risk of duplicate implementations or missed abstractions
- Files created in the wrong place because the agent didn't know what already existed

## Proposal

A **Codebase Map** — a single `.gsd/CODEBASE.md` file that gives the agent a table of contents for the entire project: what exists, where it lives, and what it does.

### Core design (simplified from original proposal)

- **Single file**: `.gsd/CODEBASE.md` — agent-readable, agent-writable
- **Generation**: Walk `git ls-files`, group by directory tree, output with one-liner descriptions
- **Maintenance**: Agent updates descriptions as it creates/modifies files during normal workflow
- **Integration**: Read at session start for orientation, regenerate at slice end to stay current
- **Token budget**: ~2-4K tokens for 100 files, scaling to ~20K for 600 files

### Implementation approach

Build as a **GSD extension** (~100-120 lines) that:
- Auto-runs on `complete-slice` (or via post-unit hook)
- Exposes `describe` as a tool the agent can call during execution
- Respects `.gitignore` via `git ls-files`
- Supports configurable directory/pattern exclusions

### What we're NOT building

The original proposal included several features that add complexity without proportional value:
- **Custom YAML data file + generated markdown** — one file is sufficient
- **Git blob hash staleness tracking** — descriptions stay current naturally if the agent updates them as it works
- **Removed files audit trail** — git history serves this purpose
- **7-subcommand CLI** — the extension pattern is simpler

## Credit

Proposed by @quique on Discord with a working draft implementation (~550 lines TypeScript). The concept is solid and addresses a real gap in GSD's project awareness. This issue tracks a simplified native integration.

---

*Community proposal — see Discord for full discussion and original implementation.*

### PR DESCRIPTION
**Pull request by jeremymcs**
## TL;DR

**What:** `/gsd codebase` command that generates `.gsd/CODEBASE.md` — a structural table of contents for the project.
**Why:** Fresh agent contexts pay a 10-30+ tool call "exploration tax" to understand what exists and where things live.
**How:** Walks `git ls-files`, groups by directory, renders with one-liner descriptions. Injected into system prompt at session start.

## What

- `codebase-generator.ts` — walks git ls-files, groups by directory, renders with one-liner descriptions, supports incremental updates that preserve existing descriptions
- `commands-codebase.ts` — CLI handler (generate, update, stats)
- `system-context.ts` — injects CODEBASE.md into system prompt at session start (alongside KNOWLEDGE.md)
- `paths.ts` — adds CODEBASE to GSD_ROOT_FILES
- `catalog.ts` — registers command with nested completions

**Files:** 7 files changed, 698 insertions.

## Why

Every fresh agent context starts blind. Without structural awareness, the agent burns 10-30+ tool calls just mapping the codebase before it can do useful work. CODEBASE.md gives instant orientation.

Closes #2229

## How

Generator walks `git ls-files`, groups files by directory, and renders a markdown document with directory headings and file listings. Directories with >20 files are collapsed to a summary. Incremental update mode preserves agent-written descriptions while adding new files.

Token budget: ~2-4K for 100 files, scaling to ~20K for 500. Configurable excludes and max file count.

### Change type
- [x] `feat` — New feature or capability
- [x] `test` — 14 unit tests

### AI-assisted contribution
This PR was developed with AI assistance (GSD/Claude). 14 unit tests, all passing.

### CODE REVIEW

File: src/resources/extensions/gsd/codebase-generator.ts [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+  for (const pattern of excludes) {
+    if (pattern.endsWith("/")) {
+      if (filePath.startsWith(pattern) || filePath.includes(`/${pattern}`)) return true;
+    } else if (filePath === pattern || filePath.endsWith(`/${pattern}`)) {
+      return true;
+    }
+  }
+  // Skip binary/lock files
+  const ext = extname(filePath).toLowerCase();
+  if ([".lock", ".png", ".jpg", ".jpeg", ".gif", ".ico", ".woff", ".woff2", ".ttf", ".eot", ".svg"].includes(ext)) {
```
Comment (igouss) [inline src/resources/extensions/gsd/codebase-generator.ts:107]: this should be project specific too.

Comment (trek-e) [reply src/resources/extensions/gsd/codebase-generator.ts:107]: Good point. If you have a specific suggestion for how to scope this per-project, that would be helpful for the author to act on.

Comment (0mm-mark) [reply src/resources/extensions/gsd/codebase-generator.ts:107]: > > this should be project specific too.
> > 
> Good point. If you have a specific suggestion for how to scope this per-project, that would be helpful for the author to act on.

If you are willing to use prior art, you can avoid a whole world of (token tax) pain....

Just include the `pi-rustdex` package:

```
rustdex semantic --help
Semantic search using natural language

Usage: rustdex semantic [OPTIONS] <QUERY>

Arguments:
  <QUERY>  Natural language query

Options:
  -r, --repo <REPO>    Optional repo name
  -l, --limit <LIMIT>  Max number of results (default 10) [default: 10]
      --json           Output results as JSON
  -h, --help           Print help
```
or in case you have specifics in mind:
```
❯ rustdex search --help
Search for symbols by name

Usage: rustdex search [OPTIONS] <QUERY>

Arguments:
  <QUERY>  The symbol name to search for

Options:
  -r, --repo <REPO>    Optional repo name to limit search
  -k, --kind <KIND>    Optional kind to filter (e.g., function, class)
  -l, --limit <LIMIT>  Max number of results (default 20) [default: 20]
      --json           Output results as JSON
  -h, --help           Print help
```

**Review by trek-e — COMMENTED** [review]
## Re-review: Codebase map

New commit: `fix(gsd): harden codebase-map -- bug fixes, UX polish, and expanded tests`. This addresses robustness concerns for a feature of this scope. The hardening commit suggests the initial implementation had edge cases discovered in testing, which is expected for a codebase analysis feature.

No new concerns from my side. The external comment from @0mm-mark should be reviewed for any actionable feedback.
